File: JavaScript/11. Container With Most Water.js
// noinspection JSUnusedGlobalSymbols
/**
 * @param {number[]} height
 * @return {number}
 */
const maxArea = (height) => {
  let start = 0
  let end = height.length - 1
  let max = 0
  const highest = Math.max(...height)

  while (end * highest > max) {
    const left = height[start]
    const right = height[end]

    const min = Math.min(left, right)
    max = Math.max(max, min * (end - start))

    if (left < right) {
      start++
    } else {
      end--
    }
  }

  return max
}

File: C-Sharp/11. Container With Most Water.cs
using System.Diagnostics.CodeAnalysis;

namespace LeetCode.C_Sharp;

[SuppressMessage("ReSharper", "UnusedType.Global")]
[SuppressMessage("ReSharper", "UnusedMember.Global")]
public partial class Solution
{
    public int MaxArea(int[] height)
    {
        int start = 0;
        int end = height.Length - 1;
        int max = 0;

        while (end > start)
        {
            int left = height[start];
            int right = height[end];
            int min = Math.Min(left, right);
            max = Math.Max(max, min * end - start);

            if (left < right)
            {
                start++;
            }
            else
            {
                end--;
            }
        }
        return max;
    }
}
